moving SRA tests to their own test group

lbergelson opened the pull request:
removing SRA tests from `ant test` and moving them to their own `ant sra-test` target

   this should isolate SRA tests so that they can no longer fail during a picard release
   
once #461 and  #442 are fixed they should be moved back into the standard `ant test`

- added a new ant target sra-test which exclusively runs the sra tests
- made all SRA test classes descend from a new `AbstractSRATest` base class
  - this ensures they are all set in the appropriate test group
  - each SRA test method will run `assertSRAIsSupported()` and be skipped if it is not
    - removed the newly redundant checks in every SRA test
 - pulled `checkAlignedUnalignedCountsByIterator` up into the base class to remove duplicated code

- updated .travis.yml to run sra-test after the rest of the tests

fixes #436 and #459

lbergelson commented:
@droazen I know you're swamped, but I was hoping you could suggest how to avoid the big copy/paste in the build.xml.  I don't know how you get ant tasks to inherit from one another.

nh13 commented on build.xml:
```diff
@@ -50,10 +50,10 @@
     </target>
     <target name="set_excluded_test_groups_non_unix" unless="isUnix">
-        <property name="excludedTestGroups" value="slow, broken, unix"/>
+        <property name="excludedTestGroups" value="slow, broken, unix, SRA"/>
```
the other groups are lowercase, would it make sense to do the same here?

lbergelson replied:
lower cased

nh13 commented on src/tests/java/htsjdk/samtools/sra/AbstractSRATest.java:
```diff
@@ -0,0 +1,56 @@
+                                                              int numberAlignments, int numberUnalignments) {
+        int countAlignments = 0, countUnalignments = 0;
+        while (
+                true) {
```
looks like there's an errant newline

lbergelson replied:
done

nh13 commented on src/tests/java/htsjdk/samtools/sra/AbstractSRATest.java:
```diff
@@ -0,0 +1,56 @@
+
+    protected static void checkAlignedUnalignedCountsByIterator(SAMRecordIterator samRecordIterator,
+                                                              int numberAlignments, int numberUnalignments) {
+        int countAlignments = 0, countUnalignments = 0;
```
while we're here, can we change these?
```
numberAlignments   -> expectedNumMapped
numberUnalignments -> expectedNumUnmapped
countAlignments    -> numMapped
countUnalignments  -> numUnmapped
```

nh13 replied:
If you change them here, can you change them in the tests too?

nh13 commented on src/tests/java/htsjdk/samtools/sra/AbstractSRATest.java:
```diff
@@ -0,0 +1,56 @@
+            try {
+                record = samRecordIterator.next();
+                Assert.assertTrue(hasRecord); // exception is not thrown if we came to this point
+            } catch (NoSuchElementException e) {
```
can be `final`

lbergelson replied:
done

nh13 commented on src/tests/java/htsjdk/samtools/sra/AbstractSRATest.java:
```diff
@@ -0,0 +1,56 @@
+                Assert.assertFalse(hasRecord);
+            }
+
+            Assert.assertEquals(hasRecord, record != null);
```
Is this needed given we check `hasRecord` on line 31 and 33?

lbergelson replied:
removed, seems redundant

lbergelson replied:
Adding an explicit non-null check for record as well.

nh13 commented on src/tests/java/htsjdk/samtools/sra/AbstractSRATest.java:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+
+    protected static void checkAlignedUnalignedCountsByIterator(SAMRecordIterator samRecordIterator,
```
Could you rename this while we are making changes?  Perhaps `checkMappedStatusCounts`?

lbergelson replied:
going with `assertCorrectCountsOfMappedAndUnmappedRecords` verbose but hopefully clear

nh13 commented on src/tests/java/htsjdk/samtools/sra/AbstractSRATest.java:
```diff
@@ -0,0 +1,56 @@
+            }
+        }
+
+        if (numberAlignments != -1) {
```
Could you add doc stating that if this is `-1` that the values wont be checked, only `hasRecord`?

lbergelson replied:
fixed

nh13 commented on src/tests/java/htsjdk/samtools/sra/AbstractSRATest.java:
```diff
@@ -0,0 +1,56 @@
+        if (numberAlignments != -1) {
+            Assert.assertEquals(numberAlignments, countAlignments);
+        }
+        if (numberUnalignments != -1) {
```
Same comment.

lbergelson replied:
fixed

nh13 commented on src/tests/java/htsjdk/samtools/sra/SRAIndexTest.java:
```diff
@@ -64,15 +62,13 @@ public void testLevelSize() {
     public void testLevelForBin() {
-        if (!SRAAccession.isSupported()) return;
-
         SRAIndex index = getIndex(DEFAULT_ACCESSION);
```
So many missing `final`s; my IntelliJ would not be pleased.

lbergelson replied:
I think I got them all.  For some reason my intellij isn't reporting them in the test files and I can't figure out why not...

nh13 commented on src/tests/java/htsjdk/samtools/sra/SRAQueryTest.java:
```diff
@@ -45,8 +39,6 @@ public void testUnmappedCounts(String acc, int numberUnalignments) {
 
     @Test(dataProvider = "testReferenceAlignedCounts")
     public void testReferenceAlignedCounts(String acc, String reference, int refernceStart, int numberAlignments) {
```
type in `refernceStart`; also, most if not all parameters in the test methods can be final.  Since we are touching so much of the code, would it make sense to add in these finishing touches?

lbergelson replied:
fixed

nh13 commented on src/tests/java/htsjdk/samtools/sra/SRAQueryTest.java:
```diff
@@ -68,8 +60,6 @@ public void testReferenceAlignedCounts(String acc, String reference, int refernc
 
     @Test(dataProvider = "testQueryCounts")
     public void testQueryCounts(String acc, String reference, int refernceStart, int referenceEnd, boolean contained, int numberAlignments, int numberUnalignment) {
```
typo here too.

lbergelson replied:
fixed

droazen commented:
@lbergelson Discussed in person ways to avoid the duplication in `build.xml` -- ping me once you've made that change, and I'll review.

lbergelson commented:
@nh13 Thanks for the review.  I believe I've addressed your concerns now.

@droazen could you give my ant file changes a glance to be sure they're sane?

droazen commented on build.xml:
```diff
@@ -38,6 +38,7 @@
     <property name="javac.target" value="1.8"/>
     <property name="javac.debug" value="true"/>
 
+    <property name="includedTestGroups" value=""/>
```
Can you make this a param in the `antcall` for the `test` target? It is confusing to have an empty value defined up here like this.

droazen commented on build.xml:
```diff
@@ -145,6 +147,20 @@
     </target>
 
+    <target name="test" depends="compile, set_excluded_test_groups" description="Run unit tests">
+        <antcall target="run-test">
```
Pass in `includedTestGroups` as a param here as well instead of relying on a property.

droazen commented on build.xml:
```diff
@@ -145,6 +147,20 @@
 
+    <target name="test" depends="compile, set_excluded_test_groups" description="Run unit tests">
+        <antcall target="run-test">
+            <param name="excludedTestGroups" value="${excludedTestGroups}, sra"/>
```
What if `excludedTestGroups` is empty? Then you'd get `value=", sra"`. Will this be ok?

droazen replied:
Never mind, I see that `excludedTestGroups` is guaranteed to be set to a non-empty value elsewhere in the build.

droazen commented on build.xml:
```diff
@@ -145,6 +147,20 @@
+    <target name="sra-test" depends="compile, set_excluded_test_groups" description="Run SRA unit tests">
+        <antcall target="run-test">
+            <param name="includedTestGroups" value="sra"/>
+            <param name="excludedTestGroups" value="${excludedTestGroups}"/>
```
The params in an `antcall` will fail to override properties that are explicitly set on the command line, as I understand things. Are these include/exclude properties ever explicitly set on the command line (using `-D`)?

droazen commented on build.xml:
```diff
@@ -119,10 +120,11 @@
 
     <!-- TEST -->
-    <target name="test" depends="compile, set_excluded_test_groups" description="Run unit tests">
+    <target name="run-test" depends="compile, set_excluded_test_groups" description="Run unit tests">
```
I think you might want to turn `run-test` into a macro using `macrodef` (there are examples of this elsewhere in this build file). Macros can be parameterized a bit more cleanly (without using `antcall`).

droazen commented:
@lbergelson I took a look, and my main suggestion is to turn `run-test` into a macro. Macros can be parameterized using attributes (no `antcall` needed), and you don't have to worry about potential collision issues with properties set explicitly on the command line, etc.

tfenne commented:
@lbergelson: I'm not trying to be a pain, and I wish I'd thought to say this sooner but would it not be less work to wrap up your conversion to gradle (and test the Picard build), than to invest all this time with the ant build?

lbergelson commented:
@tfenne That's a good point.... this would be trivial in gradle.

lbergelson commented:
I sort of expected it to be trivial in ant too... but I've already had to learn more about ant than I wanted to know.

tfenne commented:
Yeah - Ant's 100% declarative model makes some things way way more difficult than they need to be :(

droazen commented:
I think this qualifies as an urgent change that should not wait for the gradle migration (it's causing major headaches for people right now)
